docs/site/Interceptors.md

```diff
@@ -0,0 +1,552 @@
+---
+lang: en
+title: 'Interceptors'
+keywords: LoopBack 4.0, LoopBack 4
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Interceptors.html
+---
+
+## Overview
+
+Interceptors are reusable functions to provide aspect-oriented logic around
+method invocations. There are many use cases for interceptors, such as:
+
+- Add extra logic before / after method invocation, for example, logging or
+  measuring method invocations.
+- Validate/transform arguments
+- Validate/transform return values
+- Catch/transform errors, for example, normalize error objects
+- Override the method invocation, for example, return from cache
+
+## Basic use
+
+### Interceptors on controllers
+
+Interceptors are supported for public controller methods (including both static
+and prototype) and handler functions for REST routes.
+
+Controller methods decorated with `@intercept` are invoked with applied
+interceptors for corresponding routes upon API requests.
+
+```ts
+import {intercept} from '@loopback/context';
+
+@intercept(log)
+export class OrderController {
+  @intercept('caching-interceptor')
+  async listOrders(userId: string) {
+    // ...
+  }
+}
+```
+
+It's also possible to configure global interceptors that are invoked before
+method level interceptors. For example, the following code registers a global
+`caching-interceptor` for all methods.
+
+```ts
+app
+  .bind('caching-interceptor')
+  .toProvider(CachingInterceptorProvider)
+  .apply(asInterceptor);
+```
+
+Global interceptors are also executed for handler functions without a controller
+class:
+
+```ts
+/**
+ * OpenAPI operation spec for `toLowerCase(text: string)`
+ */
+const toLowerCaseOperationSpec = anOperationSpec()
+  .withOperationName('toLowerCase')
+  .withParameter({
+    name: 'text',
+    in: 'path',
+    schema: {
+      type: 'string',
+    },
+  })
+  .withStringResponse()
+  .build();
+
+/**
+ * A plain function to convert `text` to lower case
+ * @param text
+ */
+function toLowerCase(text: string) {
+  return text.toLowerCase();
+}
+
+app.route('get', '/toLowerCase/{text}', toLowerCaseOperationSpec, toLowerCase);
+```
+
+### Create a proxy to apply interceptors
+
+A
+[proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)
+can be created from the target class or object to apply interceptors. This is
+useful for the case that a controller declares dependencies of repositories or
+services and would like to allow repository or service methods to be
+intercepted.
+
+```ts
+import {createProxyWithInterceptors} from '@loopback/context';
+
+const proxy = createProxyWithInterceptors(controllerWithClassInterceptors, ctx);
+const msg = await proxy.greet('John');
+```
+
+There is also an `asProxyWithInterceptors` option for binding resolution or
+dependency injection to return a proxy to apply interceptors as methods are
+invoked.
+
+```ts
+class DummyController {
+  constructor(
+    @inject('my-controller', {asProxyWithInterceptors: true})
```

Comment:
The addition of proxy support is making this pull request unnecessarily large. Can you please remove proxies from the first iteration, to make it easier for us to review the changes? At the moment, this pull request has 1800 new lines of code :(

Reply:
The addition is in its own commit. Would that help?

Reply:
Not really, but I'll do my best.